dropshot/src/api_description.rs

```diff
@@ -919,27 +925,59 @@ impl<Context: ServerContext> ApiDescription<Context> {
                 }
             };
 
+            // If the endpoint defines an error type, emit that for
```

Comment:
So, I'd like to put them in `components.responses`, too. The reason I didn't is that it might be a bit annoying to determine the name for each response schema. `schemars` internally disambiguates colliding schema names by turning subsequent ones into like `Error2` or whatever, but (AFAICT) we only get that when we actually generate the schema and it gives us back a reference (into `components.schemas`). We could then try to parse that reference and get the name back out to then use it to generate a `components.responses` entry for that response, which seems possible, I just thought it seemed annoying enough that I didn't *really* want to bother with it. Do you think it's worth doing?

Reply:
I don't believe that deduplication is applied to `JsonSchema::schema_name()`; as
far as I can tell, it only happens once a schema has already been generated,
because that's when the generator can check if the name already exists in the
set of schemas that have been generated so far?

Reply:
Ah. Unfortunately, it turns out that Progenitor chokes on outputting two separate response objects that have the same response body schema; it currently expects there to be exactly one response object for errors, as the current code hits [this assertion](https://github.com/oxidecomputer/progenitor/blob/96bdb54c12f5ba02b6ca5f9213083e8d9c4b1e4f/progenitor-impl/src/method.rs#L1277-L1279).

I'm on the fence about whether the right solution here is to change the OpenAPI spec generation in Dropshot to emit a single response schema object for 4xx and 5xx errors, or to make Progenitor smart enough to understand that multiple response schemas can share the same _response body type schema_, and recognize that those are the same thing. What do you think @ahl?

Reply:
What's the OpenAPI input it's choking on? I wouldn't have thought that this move away from `components/responses` would have had this result.

With regard to progenitor supporting it; I would like to do that, but I would also love to be able to defer that. It's not a small amount of work to do reasonably well.

Reply:
Progenitor panics when building oxidecomputer/omicron#7196, which updates Omicron to depend on this branch. Note that the only change to the schemas is to generate separate response schemas for 4xx and 5xxs, as that PR hasn't added any user-defined error types or anything. It'll die on any of the OpenAPI docs in that PR, which one it hits first seems to be non-deterministic and depends on the order in which crates are built. For an example of the OpenAPI change, see https://github.com/oxidecomputer/omicron/pull/7196/files#diff-a389634108eedc06638f772e1384ef397e4d4254363b3459e5043341227f06b6.

I'm going to take another crack at having Dropshot generate one response schema for both error status codes, which would hopefully resolve the issue without requiring a change to Progenitor.

Reply:
I'm genuinely surprised that progenitor is dying despite its constraints. My recollection was that a reference to a response would be handled identically to an inline response. I'm going to look on the progenitor side.

Reply:
Okay @ahl, if you look at the PR comparison view again as of commit f0e6a46acbcb21774cc50ba24ee2faa0b9e060a3, you'll see that there's now *no diff* in any of the generated OpenAPI documents compared to `main`; the only diff in the expectorate output now is the new schemas with user-defined error types!  

So, while it would be nice for progenitor to be able to handle cases like this, APIs that don't use custom error types will now emit backwards-compatible OpenAPI documents that the current version of Progenitor ought to be able to handle correctly.